java/ql/src/Violations of Best Practice/Undesirable Calls/DoNotCallFinalize.md

```diff
@@ -0,0 +1,60 @@
+## Overview
+
+Triggering garbage collection by directly calling `finalize()` may either have no effect or may trigger unnecessary garbage collection, leading to erratic behavior, performance issues, or deadlock.
+
+## Recommendation
+
+Avoid calling `finalize()` in application code. Allow the JVM to determine a garbage collection schedule instead. If you need to explicitly release resources, provide a specific method to do so, such as by implementing the `AutoCloseable` interface and overriding its `close` method. You can then use a `try-with-resources` block to ensure that the resource is closed.
+
+## Example
+
+```java
+class LocalCache {
+    private Collection<File> cacheFiles = ...;
+    // ...
+}
+
+void main() {
+    LocalCache cache = new LocalCache();
+    // ...
+    cache.finalize(); // NON_COMPLIANT
+}
+
+```
+
+```java
+import java.lang.AutoCloseable;
+import java.lang.Override;
+
+class LocalCache implements AutoCloseable {
+    private Collection<File> cacheFiles = ...;
+    // ...
+
+    @Override
+    public void close() throws Exception {
+        // release resources here if required
+    }
+}
+
+void main() {
+    // COMPLIANT: uses try-with-resources to ensure that
+    // a resource implementing AutoCloseable is closed.
+    try (LocalCache cache = new LocalCache()) {
+        // ...
+    }
+}
+
+```
+
+# Implementation Notes
+
+This rule ignores `super.finalize()` calls that occur within `finalize()` overrides since calling the superclass finalizer is required when overriding `finalize()`. Also, although overriding `finalize()` is not recommended, this rule only alerts on direct calls to `finalize()` and does not alert on overrides of `finalize()`.
```

Comment:
You know, I said this was clear, but I think this would make it clearer.
```suggestion
This rule ignores `super.finalize()` calls that occur within `finalize()` overrides since calling the superclass finalizer is required when overriding `finalize()`. Also, although overriding `finalize()` is not recommended, this rule only alerts on direct calls to `finalize()` and does not alert on calls to overrides of `finalize()`.
```

Reply:
Hmm, I don't mean it doesn't alert on calls to overrides of `finalize()`. That's exactly what it does alert on, so maybe I need to rewrite this a bit more if that wasn't clear. 😅  What I mean is it does not alert on the override implementation itself, i.e. it does not alert on [line 12](https://github.com/github/codeql/pull/19075/files#diff-5b9ede1bbe893ac873b9fccb9050da84987ed2fd35171c30612c1e9ebc860b4dR12) in the tests; whereas, `java/overriden-or-overloaded-finalize` will alert on line 12.

Reply:
Is "does not alert on **method** overrides of `finalize()`" clearer? Updated to that in [0380279](https://github.com/github/codeql/pull/19075/commits/0380279c39079c713a1d64a412669b18abccee7a).

Reply:
I think a wording that would make it clearer is "method declarations overriding `finalize()`". (Or "definition" - in C/C++ they're quite separate concepts but in java I think the only time you have a declaration without a definition is interface methods and abstract methods.)

Reply:
Updated in [fc21abc](https://github.com/github/codeql/pull/19075/commits/fc21abc7e4af6b6b1045c5b3e25e8a7114126680).